formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt

```diff
@@ -294,36 +303,54 @@ class Cbor(val updateMode: UpdateMode = UpdateMode.BANNED, val encodeDefaults: B
                 return if (negative) -(value + 1).toLong()
                 else value.toLong()
             }
-            val buf = input.readToByteBuffer(bytesToRead)
-            val res = when (bytesToRead) {
-                1 -> buf.getUnsignedByte().toLong()
-                2 -> buf.getUnsignedShort().toLong()
-                4 -> buf.getUnsignedInt()
-                8 -> buf.getLong()
-                else -> throw AssertionError()
-            }
+            val res = input.readExact(bytesToRead)
             return if (negative) -(res + 1)
             else res
         }
 
+        private fun InputStream.readExact(bytes: Int): Long {
+            val arr = readExactNBytes(bytes)
+            var result = 0L
+            for (i in 0 until bytes) {
+                result = (result shl 8) or (arr[i].toInt() and 0xFF).toLong()
+            }
+            return result
+        }
+
         fun nextFloat(): Float {
             if (curByte != NEXT_FLOAT) throw CborDecodingException("float header", curByte)
-            val res = input.readToByteBuffer(4).getFloat()
+            val res = Float.fromBits(readInt())
             readByte()
             return res
         }
 
         fun nextDouble(): Double {
             if (curByte != NEXT_DOUBLE) throw CborDecodingException("double header", curByte)
-            val res = input.readToByteBuffer(8).getDouble()
+            val res = Double.fromBits(readLong())
             readByte()
             return res
         }
 
+        private fun readLong(): Long {
+            var result = 0L
+            for (i in 0..7) {
+                val byte = input.read()
+                result = (result shl 8) or byte.toLong()
```

Comment:
Won't `byte.toLong` make the sign-extending conversion? It should be `byte.toLong() and 0xFFL`, probably.

Reply:
No, because `InputStream.read` returns `Int` type with a values within -1..255 range